src/rust/src/wallet.rs

```diff
@@ -130,38 +149,159 @@ pub enum WalletError {
         }
+        self.witness_tree = BridgeTree::new(MAX_CHECKPOINTS);
+        self.last_checkpoint = None;
+        self.last_observed = None;
+        self.locked_notes = BTreeSet::new();
+        self.spent_notes = BTreeSet::new();
     }
 
-    pub fn checkpoint_witness_tree(&mut self) {
+    /// Checkpoints the note commitment tree. This returns `false` and leaves the note
+    /// commitment tree unmodified if the block height does not immediately succeed
+    /// the last checkpointed block height (unless the note commitment tree is empty,
+    /// in which case it unconditionally succeeds). This must be called exactly once
+    /// per block.
+    pub fn checkpoint(&mut self, block_height: BlockHeight) -> bool {
+        // checkpoints must be in order of sequential block height and every
+        // block must be checkpointed
+        if self
+            .last_checkpoint
+            .iter()
+            .any(|last_height| block_height != *last_height + 1)
+        {
+            return false;
+        }
+
         self.witness_tree.checkpoint();
+        self.last_checkpoint = Some(block_height);
+        true
+    }
+
+    /// Returns whether or not a checkpoint has been created. If no checkpoint exists,
+    /// the wallet has not yet observed any blocks.
+    pub fn is_checkpointed(&self) -> bool {
+        self.last_checkpoint.is_some()
     }
 
-    pub fn rewind_witness_tree(&mut self) -> bool {
-        self.witness_tree.rewind()
+    /// Rewinds the note commitment tree to the given height.
+    ///
+    /// Returns `false` if the rewind would exceed the number of stored checkpoints.
+    pub fn rewind(&mut self, to_height: BlockHeight) -> Result<u32, RewindError> {
+        if let Some(checkpoint_height) = self.last_checkpoint {
+            if to_height > checkpoint_height {
+                return Err(RewindError::NoRewindNeeded(checkpoint_height));
+            }
+
+            let blocks_to_rewind = <u32>::from(checkpoint_height) - <u32>::from(to_height);
+
+            // if the rewind length exceeds the number of checkpoints we have stored,
+            // return failure.
+            let checkpoint_count = self.witness_tree.checkpoints().len();
+            if blocks_to_rewind as usize > checkpoint_count {
+                return Err(RewindError::InsufficientCheckpoints(checkpoint_count));
+            }
+
+            let mut blocks_rewound: u32 = 0;
+            while blocks_rewound < blocks_to_rewind {
+                // we've already checked that we have enough checkpoints to fully
+                // rewind by the requested amount.
+                assert!(self.witness_tree.rewind());
+                blocks_rewound += 1;
+            }
+
+            // reset our last observed height to ensure that notes added in the future are
+            // from a new block
+            let last_observed = LastObserved {
+                block_height: checkpoint_height - blocks_rewound,
+                block_tx_idx: None,
```

Comment:
What about `to_height == checkpoint_height` in this function? Shouldn't we no-op? (Otherwise, among maybe other things, `block_tx_idx` here gets needlessly set to `None`.)

Reply:
I set `block_tx_idx` to `None` here because the information in it will be invalid, and I don't want to leave some stale index in there that could cause problems at any later point.

Reply:
Now that I look at this more closely, I could change the `to_height > checkpoint_height` condition to just be `>=` and then return `Ok(0)` in that case; that would provide the no-op that you're requesting here. That would get rid of the `NoRewindNeeded` case.